### PR TITLE
Contract and Harnesses for `<*const T>::offset_from`

### PR DESCRIPTION
**Pull request by stogaru**
Towards #76 

### Changes

* Adds contracts for `<*const T>::offset_from`.
* Adds harnesses for the function verifying:
    * All integer types
    * Tuples (composite types)
    * Unit Type
* Accomplishes this using a macro.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Review by celinval — COMMENTED** [review]
Thank you!

File: library/core/src/ptr/const_ptr.rs [hunk 1]
```diff
@@ -667,6 +671,13 @@ impl<T: ?Sized> *const T {
     #[rustc_const_stable(feature = "const_ptr_offset_from", since = "1.65.0")]
     #[inline]
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
+    #[requires(
+        (mem::size_of::<T>() != 0) && // offset_from requires pointee size to be greater than 0
+        (self as isize).checked_sub(origin as isize).is_some() &&
+        (self as isize - origin as isize) % (mem::size_of::<T>() as isize) == 0 &&
+        kani::mem::same_allocation(self, origin)
```
[the remark this answers is not in the document]
Comment (stogaru) [reply library/core/src/ptr/const_ptr.rs:678]: @celinval Wouldn't same_allocation consider that case? In other words, if they are pointing to the same address, can they be of different provenance?

[the remark this answers is not in the document]
Comment (celinval) [reply library/core/src/ptr/const_ptr.rs:678]: I believe `same_allocation` requires that at least one pointer is allocated.

Unfortunately, Kani doesn't quite model provenance today like the Rust specification. In Rust, you can have pointers that point to the same address that have different provenance. For example, pointer to an object and the pointer to its first field.

If you run the following example with MIRI, you will see that UB is detected:
```rust
use std::ptr::addr_of_mut;

#[derive(Clone, Copy, Default)]
#[repr(C)]
struct Point {
    x: usize,
    y: usize,
}
fn main() {
    let mut point = Point::default();
    let ptr = addr_of_mut!(point);
    let x_ptr = addr_of_mut!(point.x);
    assert_eq!(ptr as usize, x_ptr as usize);
    
    // Not safe!
    let y_ptr = unsafe { x_ptr.add(1) };
    unsafe { *y_ptr = 10 };
    assert_eq!(point.y, 10);
}
```

This is the error:
```
error: Undefined Behavior: attempting a write access using <1569> at alloc763[0x8], but that tag does not exist in the borrow stack for this location
  --> src/main.rs:17:14
   |
17 |     unsafe { *y_ptr = 10 };
   |              ^^^^^^^^^^^
   |              |
   |              attempting a write access using <1569> at alloc763[0x8], but that tag does not exist in the borrow stack for this location
   |              this error occurs as part of an access at alloc763[0x8..0x10]
   |
   = help: this indicates a potential bug in the program: it performed an invalid operation, but the Stacked Borrows rules it violated are still experimental
   = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
help: <1569> was created by a SharedReadWrite retag at offsets [0x0..0x8]
  --> src/main.rs:12:17
   |
12 |     let x_ptr = addr_of_mut!(point.x);
   |                 ^^^^^^^^^^^^^^^^^^^^^
   = note: BACKTRACE (of the first span):
   = note: inside `main` at src/main.rs:17:14: 17:25
   = note: this error originates in the macro `addr_of_mut` (in Nightly builds, run with -Z macro-backtrace for more info)

```

File: library/core/src/ptr/const_ptr.rs [hunk 2]
```diff
@@ -1899,3 +1910,69 @@ impl<T: ?Sized> PartialOrd for *const T {
...
+
+    macro_rules! generate_offset_from_harness {
+        ($type: ty, $proof_name: ident) => {
+            #[kani::proof_for_contract(<*const $type>::offset_from)]
+            pub fn $proof_name() {
+                let val: $type = kani::any::<$type>();
+                let ptr: *const $type = &val;
+
+                // offset the src pointer within the allocation bounds non-deterministically
+                let offset: usize = kani::any_where(|x| *x <= mem::size_of::<$type>());
```
[the remark this answers is not in the document]
Comment (stogaru) [reply library/core/src/ptr/const_ptr.rs:1938]: Our challenge requires us to verify both integer types and slices. So, would verifying a slice of integer types satisfy the requirement of verifying both slices and integer pointee types?

Also for verifying slices, should we generate pointers using the PointerGenerator or by getting a raw pointer non-deterministically offsetted within a slice generated using `kani::Arbitrary::any_array`?

File: library/core/src/ptr/const_ptr.rs [hunk 3]
```diff
@@ -667,6 +671,13 @@ impl<T: ?Sized> *const T {
     #[rustc_const_stable(feature = "const_ptr_offset_from", since = "1.65.0")]
     #[inline]
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
+    #[requires(
+        (mem::size_of::<T>() != 0) && // offset_from requires pointee size to be greater than 0
+        (self as isize).checked_sub(origin as isize).is_some() &&
+        (self as isize - origin as isize) % (mem::size_of::<T>() as isize) == 0 &&
```
[the remark this answers is not in the document]
Comment (stogaru) [reply library/core/src/ptr/const_ptr.rs:677]: But self and origin need not be aligned right? It is only distance between them that needs to be a multiple of `mem::size_of::<T>()`.

File: library/core/src/ptr/const_ptr.rs [hunk 4]
```diff
@@ -1899,3 +1910,69 @@ impl<T: ?Sized> PartialOrd for *const T {
         *self >= *other
     }
 }
+
+#[cfg(kani)]
+#[unstable(feature = "kani", issue = "none")]
+mod verify {
```
[the remark this answers is not in the document]
Comment (stogaru) [reply library/core/src/ptr/const_ptr.rs:1919]: Used PointerGenerator to cover dangling pointers.

**Review by celinval — APPROVED** [review]

**Review by carolynzech — APPROVED** [review]

**Review by feliperodri — APPROVED** [review]

**Pull request merged**